Add a performance guide. (#442)
bytewax/bytewax: docs/guide/advanced-concepts/performance.md

As with all performance tuning, it is important to have good tooling to measure the
effect of any changes you make. See the section on <project:#xref-profiling>
for some tips on how to profile Bytewax.

One of the worthwhile trade-offs of dataflow programming is having to play
by the strict rules of operators and state management. In return you gain
the ability to more-easily scale your dataflow across a cluster for increased
performance. Unfortunately "more-easily" does not mean "trivially": Practically
speaking, only adding more workers to a dataflow does not automatically result
in an improvement in performance. The reality of all performance optimization
is matching computational resources to the specific needs of your computation,
and in this document we'll describe concrete design recommendations you should
follow while implementing your dataflow to play to the strengths of Bytewax and
ensure your dataflow is performant.

## Bytewax's worker execution model

A Bytewax **worker** is a thread that is executing your dataflow. Workers can be grouped
into separate processes, but refer to a single thread within a single process.

> Bytewax's execution model uses identical workers. Workers execute all of the steps

Reviewer: This might be a bit misleading, as not every worker will always execute every step, I'd phrase it a bit differently:

```suggestion
Bytewax's execution model uses identical workers. Workers can execute any of the steps
```

in a dataflow, including input and output steps. Workers **may** exchange
data to ensure the semantics of operators, but do not automatically
exchange work to other workers in order to increase performance.

## Workers, processes and GIL contention

When multiple workers are started within the same Python process using
`-w/--workers-per-process`, they must contend over Python's GIL, or
{external+python:std:term}`global interpreter lock`. Python's GIL ensures
that only one Python thread can run at a time, preventing race conditions and
ensuring thread-safety.

> In practice this means that when you are attempting to increase the parallelism
> of a dataflow, should try starting with multiple processes using the
> `-i/--process-id`, `-a/--addresses` instead of multiple workers with `-w`.

Reviewer: ```suggestion
In practice this means that when you are attempting to increase the parallelism
of a dataflow, you should try starting with multiple processes using the
`-i/--process-id`, `-a/--addresses` instead of multiple workers with `-w`.
```

> If your code can make use of libraries that make progress without holding the GIL,
> adding workers to each process with `-w` may increase performance.

Reviewer: Since it's always hard to know what libraries release the GIL, I'd add an example of a part of the standard library that does:

```suggestion
If your code can make use of libraries that make progress without holding the GIL
(like python's urllib), adding workers to each process with `-w` may increase performance.
```
